Allow recursive references in single file specs aggregation.

macisamuele opened the pull request:
With the last release was added the ``dereference_served_schema`` parameter.
As described by the documentation the parameter is not working well in case of recursive model definition.

The PR includes the possibility to generate a single file swagger specs even in case of recursive model definition. It reaches the goal allowing local references only for the [definition objects](http://swagger.io/specification/#definitionsObject).

The generated specs are tested respect to the [bravado](https://github.com/Yelp/bravado) client.

Ping: @striglia, @lucagiovagnoli, @sjaensch

striglia commented on pyramid_swagger/api.py:
```diff
@@ -101,26 +101,206 @@ def view_for_api_declaration(request):
+
+def _relpath(path, start):
+    """Return a relative version of a path
+    NOTE: Code duplicated from Python 2.7.9 implementation because the
```
Interesting...can we at least link to a bug report for context?

macisamuele replied:
I discovered it running the tests.
Debugging the application I saw a different code between Python 2.6.9 and 2.7.9 and 3.4.2.
In the next days I'll look for more deeper information and/or the reason that lead Python to change the implementation of the function.

macisamuele replied:
I have found the [python issue #5117](http://bugs.python.org/issue5117) that seems explaining the difference in the ``os.path.relpath`` execution with Python 2.6.9, 2.7.9, 3.4.2 and 3.5.2.
A more direct approach could be to *practically* seen the problem. You can do it running the script that is shown below (are added also the results that I have on my laptop).

I have not found an *official* reply to your doubt; but I'm able to provide you a short snippet that highlights the issue.

Script
```bash
for version in 2.6.9 2.7.9 3.4.2 3.5.2; do
    pyenv global $version
    python -c "import os; import sys; print(sys.version); print(os.path.relpath('/dir1/../dir2/file.json', '/'))"
    echo "=========="
done
```

My Results
```bash
2.6.9 (unknown, Jul  1 2016, 10:58:34) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
../dir2/file.json
==========
2.7.9 (default, Jul  1 2016, 10:52:19) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
dir2/file.json
==========
3.4.2 (default, Jul  1 2016, 10:56:36) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
dir2/file.json
==========
3.5.2 (default, Jul  3 2016, 11:59:24) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
dir2/file.json
```
**NOTE**: I'm using pyenv to quickly switch between python version. Each command execution shows the used version and the result of the ``os.path.relpath('/dir1/../dir2/file.json', '/')`` command.

striglia replied:
Frankly, I'm inclined to not work around a 2.6-specific bug that only shows up for some uses of relpath. Does it break an existing test? Even if so, it might be reasonable to just omit 2.6 for that test and target deprecation of it in version 3.0.

striglia commented on pyramid_swagger/api.py:
```diff
@@ -101,26 +101,206 @@ def view_for_api_declaration(request):
+
+
+def _get_target(spec, target, current_path=''):
+    # Note: we assume that swagger.json file is always a server-local file
```
If target is always a local file, is there a better name for this? I'm not sure what this function does by its current naming...perhaps a docstring would help explain too?

striglia commented on pyramid_swagger/api.py:
```diff
@@ -101,26 +101,206 @@ def view_for_api_declaration(request):
+        )
+
+
+def is_a_definition(json_path):
```
What do we mean by "definition" here? If it's a swagger term, maybe we can link out to the relevant terminology in their docs? This spec unpacking logic is getting quite complicated, so I'm interested in making sure it is clear when reading through...

macisamuele replied:
The general idea was: we have troubles if we are dereferencing recursive structure. Since the only possible recursive structure are the data-model we can prevent the complete dereferencing for those objects.
Every reference to a data model should be dereferenced in the [swagger definitions object](http://swagger.io/specification/#definitionsObject). Reading the [Swagger specification](http://swagger.io/specification/) was possible to identify the objects which referees to a *definition* (as described in the comments).

Have you some suggestion about the level of detail that I should insert on the docstring? Could I put part of this comment into the method documentation.

striglia replied:
Cool. Just clarifying that you mean the Swagger definitions object in a docstring should be sufficient.

macisamuele replied:
Cool ... tomorrow morning I'll update the doc-strings and push them.

striglia commented on pyramid_swagger/api.py:
```diff
@@ -101,26 +101,206 @@ def view_for_api_declaration(request):
+                    target_url = _get_target(spec, subval, file_path)
+                    marshaled_target = _marshal_target(target_url)
+                    if marshaled_target not in defs_dict:
+                        defs_dict[marshaled_target] = None
```
Is this setting useful? Seems like the next line should make it not matter...?

macisamuele replied:
Maybe adding a comment here would have been beneficial 😄 
The placeholder (``None``) is needed because the ``resolve_ref`` could involve a recursive method call if there is a recursive data-model dereferencing and we need to have a stop condition for the recursion.

Since we are resolving the references

striglia commented on pyramid_swagger/api.py:
```diff
@@ -101,26 +101,206 @@ def view_for_api_declaration(request):
 
 
-def resolve_ref(spec, url):
+def _low_level_translate(path, is_marshaling=True):
```
With a general name like this, I think a clear docstring including precisely what type of object this returns and what the function does will be extra important.

striglia commented on pyramid_swagger/api.py:
```diff
@@ -101,26 +101,206 @@ def view_for_api_declaration(request):
+    target_scheme = target_url.scheme
+    if len(target_url.path) > 0 and \
+            target_scheme in ['', 'file', 'http', 'https']:
+        value = _low_level_translate(
```
If we're just immediately returning value, I'd do it on this line instead of the temporary variable

macisamuele replied:
Sure ... it was done only for debugging purposes

macisamuele commented:
Ping: @striglia, @lucagiovagnoli, @sjaensch
Please could you give me some feedback? In that way I can add all the required docstrings and hoping for the PR merge. Thanks a lot!

striglia commented on pyramid_swagger/api.py:
```diff
@@ -101,26 +101,268 @@ def view_for_api_declaration(request):
+           json_path[-1] in ['schema', 'items', 'allOf']
+
+
+def resolve_ref(spec, url, json_path, file_path, defs_dict):
```
Have you done some basic speed tests before and after? The additional logic hasn't made things too much slower I hope?

striglia commented:
OK, I think the major feedback is done. I'm hoping we can resolve the `relpath` issue without re-implementing it, but otherwise the tests are our standard.

Did at least one of your tests reproduce the recursive bug **and** fail before your code changes to `resolve_ref`?

macisamuele commented:
To reproduce the bug you should run the ``recursive_app_test.py`` on the master branch.
To do this you can execute the following commands from the repository directory:
    - ``git checkout origin/master``
    - ``git checkout allow-recursive-references tox.ini tests/acceptance/recursive_app_test.py``

macisamuele commented:
@striglia: From your yesterday comments I have:
 - added the DocString for the ``is_a_definition`` method
 - renamed ``is_a_definition`` to ``is_a_swagger_definition``
 - skipped some marshaller tests if using Python 2.6
 - added a note into ``docs/configuration.rst``

striglia commented:
Thanks for the patience and all the code @macisamuele  :)
